src/Compression/Compression.cs

```diff
@@ -0,0 +1,153 @@
+			if (status != CompressionStatus.Ok)
+				throw new InvalidOperationException ("Internal stream could not be created.");
+
+			// set the different required values
+			inSize = inBufferSize;
+			outSize = outBufferSize;
+			dstBuf = Marshal.AllocHGlobal (outSize);
+			srcBuf = Marshal.AllocHGlobal (inBufferSize);
+
+			internalStream.Destination = dstBuf;
+			internalStream.DestinationLength = outSize;
+			internalStream.Source = srcBuf;
+			internalStream.SourceLength = 0;
+		}
+
+		public async Task<float> Process ()
+		{
+			do {
+				if (internalStream.SourceLength == 0) {
+					if (!finalizeStream || operation == StreamOperation.Encode) {
+						// Refill source buffer
+						byte[] data = new byte[inSize];
+						await sourceStream.ReadAsync (data, 0, inSize);
+						Marshal.Copy(data, 0, internalStream.Source, data.Length);
+						totalInputSize += data.Length;
+						internalStream.Source = srcBuf;
+						internalStream.SourceLength = data.Length;
+
+						if (data.Length < inSize) {
+							// Reached end of data.
+							finalizeStream = true;
+						}
+					}
+				}
+				if (internalStream.DestinationLength == 0) {
+					// output buffer is full, copy from managed to unmanaged and write
+					byte[] data  = new byte[outSize];
+					Marshal.Copy (internalStream.Destination, data, 0, outSize);
+					await destinationStream.WriteAsync (data, 0, outSize); // we should be appending
+					totalOutputSize += outSize;
+
+					internalStream.Destination = dstBuf;
+					internalStream.DestinationLength = outSize;
+				}
+
+				// stream should be set up, perform the action
+				status = (CompressionStatus) CompressionStreamProcess (ref internalStream, 0); // TODO, use flags instead!
+				switch (status) {
+				case CompressionStatus.Ok:
+					// we will do the loop at least once more, lets be ready
+					if (internalStream.DestinationLength == 0) {
+						byte[] data  = new byte[outSize];
+						Marshal.Copy (internalStream.Destination, data, 0, outSize);
+						await destinationStream.WriteAsync (data, 0, outSize); // we should be appending
+						totalOutputSize += outSize;
+
+						internalStream.Destination = dstBuf;
+						internalStream.DestinationLength = outSize;
+					}
+					break;
+				case CompressionStatus.End:
+					// we are done, write whatever we have in the buffer
+					if ((long)internalStream.Destination > (long)dstBuf) {
+						var size = (int)((long)internalStream.Destination - (long)dstBuf);
+						byte[] data  = new byte[size];
+						Marshal.Copy (internalStream.Destination, data, 0, outSize);
+						await destinationStream.WriteAsync (data, 0, size); // we should be appending
+						totalOutputSize += size;
+					}
+					break;
+				default:
+					throw new InvalidOperationException ("An error occurred when performing the operation");
+				}
+
+			} while (status == CompressionStatus.Ok);
+
+			return totalInputSize / totalOutputSize;
+		}
+
+		protected virtual void Dispose (bool disposing)
```

Comment:
There's no finalizer, so this will only be called when manually disposing the object. You also need to add a finalizer that calls Dispose:

```csharp
~CompressionStream ()
{
    Dispose (false);
}
```